src/ssh.c

```diff
@@ -1479,137 +1476,231 @@ union wolfSSH_key {
+    byte* newKey = NULL;
+
+    WOLFSSH_UNUSED(heap);
+
+    if (*out == NULL) {
+        newKey = (byte*)WMALLOC(inSz, heap, DYNTYPE_PRIVKEY);
+        if (newKey == NULL) {
+            return WS_MEMORY_E;
+        }
+        *out = newKey;
+    }
+    else {
+        if (*outSz < inSz) {
+            WLOG(WS_LOG_DEBUG, "DER private key output size too small");
+            return WS_BUFFER_E;
+        }
+        newKey = *out;
+    }
+    *outSz = inSz;
+    WMEMCPY(newKey, in, inSz);
+
+    ret = IdentifyAsn1Key(in, inSz, 1, heap);
+    if (ret > 0) {
+        *outType = (const byte*)IdToName(ret);
+        *outTypeSz = (word32)WSTRLEN((const char*)*outType);
+        ret = WS_SUCCESS;
+    }
+
+    return ret;
+}
+
+
+static int DoPemKey(const byte* in, word32 inSz, byte** out,
+        word32* outSz, const byte** outType, word32* outTypeSz,
+        void* heap)
+{
+    int ret = WS_SUCCESS;
+    byte* newKey = NULL;
+    word32 newKeySz = inSz; /* binary will be smaller than PEM */
+
+    WOLFSSH_UNUSED(heap);
+
+    if (*out == NULL) {
+        newKey = (byte*)WMALLOC(newKeySz, heap, DYNTYPE_PRIVKEY);
+        if (newKey == NULL) {
+            return WS_MEMORY_E;
         }
-        *outSz = inSz;
-        WMEMCPY(newKey, in, inSz);
+        *out = newKey;
+    }
+    else {
+        if (*outSz < inSz) {
+            WLOG(WS_LOG_DEBUG, "PEM private key output size too small");
+            return WS_BUFFER_E;
+        }
+        newKey = *out;
+    }
 
-        ret = IdentifyKey(in, inSz, 1, heap);
+    /* If it is PEM, convert to ASN1 then process. */
+    ret = wc_KeyPemToDer(in, inSz, newKey, newKeySz, NULL);
+    if (ret > 0) {
+        newKeySz = (word32)ret;
+        ret = WS_SUCCESS;
+    }
+    else {
+        WLOG(WS_LOG_DEBUG, "Base64 decode of public key failed.");
+        ret = WS_PARSE_E;
+    }
+
+    if (ret == WS_SUCCESS) {
+        ret = IdentifyAsn1Key(newKey, newKeySz, 1, heap);
         if (ret > 0) {
+            *outSz = newKeySz;
             *outType = (const byte*)IdToName(ret);
             *outTypeSz = (word32)WSTRLEN((const char*)*outType);
             ret = WS_SUCCESS;
         }
+        else {
+            WLOG(WS_LOG_DEBUG, "unable to identify key");
+        }
     }
-    else if (format == WOLFSSH_FORMAT_PEM) {
-        word32 newKeySz = inSz; /* binary will be smaller than PEM */
 
-        if (*out == NULL) {
-            newKey = (byte*)WMALLOC(newKeySz, heap, DYNTYPE_PRIVKEY);
-            if (newKey == NULL) {
-                return WS_MEMORY_E;
-            }
-            *out = newKey;
+    return ret;
+}
+
+
+static int DoOpenSshKey(const byte* in, word32 inSz, byte** out,
+        word32* outSz, const byte** outType, word32* outTypeSz,
+        void* heap)
+{
+    int ret = WS_SUCCESS;
+    byte* newKey = NULL;
+    word32 newKeySz = inSz; /* binary will be smaller than PEM */
+
+    if (*out == NULL) {
+        newKey = (byte*)WMALLOC(newKeySz, heap, DYNTYPE_PRIVKEY);
```

Comment:
check on free of newKey with error cases

Reply:
Check.